net/tlsdial/tlsdial.go

```diff
@@ -76,17 +76,20 @@ func Config(host string, ht *health.Tracker, base *tls.Config) *tls.Config {
 	// own cert verification, as do the same work that it'd do
 	// (with the baked-in fallback root) in the VerifyConnection hook.
 	conf.InsecureSkipVerify = true
-	conf.VerifyConnection = func(cs tls.ConnectionState) error {
+	conf.VerifyConnection = func(cs tls.ConnectionState) (retErr error) {
 		// Perform some health checks on this certificate before we do
 		// any verification.
+		isSelfSigned := len(cs.PeerCertificates) > 0 && certIsSelfSigned(cs.PeerCertificates[0])
 		if ht != nil {
-			if certIsSelfSigned(cs.PeerCertificates[0]) {
-				// Self-signed certs are never valid.
-				ht.SetTLSConnectionError(cs.ServerName, fmt.Errorf("certificate is self-signed"))
-			} else {
-				// Ensure we clear any error state for this ServerName.
-				ht.SetTLSConnectionError(cs.ServerName, nil)
-			}
+			defer func() {
+				if retErr != nil && isSelfSigned {
```

Comment:
Should we log a warning if there's a self-signed certificate that _passes_ validation?

Reply:
Oh right. That's what I meant to do and why I pulled the variable up. Then I got distracted.

Reply:
PTAL